crates/sui-storage/src/fullnode_pending_tx_log.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! FullNodePendingTransactionLog is Write Ahead Log uesd in Transaction Orchestrator
+//! for transaction submission processing. It plays two roles:
+//! 1. At one time, a transaction is only processed once.
+//! 2. When Fullnode crashes and restarts, the pending transaction will be loaded and retried.
+
+use crate::mutex_table::MutexTable;
+use crate::write_ahead_log::{MUTEX_TABLE_SHARD_SIZE, MUTEX_TABLE_SIZE};
+use std::path::PathBuf;
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{TrustedTransction, VerifiedTransaction};
+use typed_store::traits::TypedStoreDebug;
+use typed_store::{rocks::DBMap, traits::Map};
+use typed_store_derive::DBMapUtils;
+
+pub type IsFirstRecord = bool;
+
+#[derive(DBMapUtils)]
+struct FullNodePendingTransactionTable {
+    logs: DBMap<TransactionDigest, TrustedTransction>,
```

Comment:
Is it important at all to capture the arrival timestamp in the log? In other words, when doing a recovery does recovery depend on which order the log should be recovered?

Reply:
good question. We have plans to only keep two write path request types: `WaitForEffectsCert` and `WaitForLocalExecution`. With these two types, a correct client (assuming only connecting to one fullnode) shouldn't have concurrent transactions with causal dependencies on fullnode. However, a buggy client could do this and make fullnode busy with retrying transactions out of order until all of them succeeds. But in this case even sorting by timestamp may not help too much? A neat solution would be something like the dependency tracker we have in `TransactionManager`, if one is encapsulated into a module.